app/models/vm.rb

```diff
@@ -125,6 +125,23 @@ def supported_consoles
     }
   end
 
+  def allow_retire_request_creation?
+    MiqRequest.where(:source_type => "Vm", :source_id => id).with_type("VmRetireRequest").find_each do |r|
+      warn_message = "MiqRequest with id:#{r.id} to retire Vm name:'#{name}' id:#{id} already created"
+      if r.request_pending_approval?
+        _log.warn("#{warn_message} but not approved yet")
+        return false
+      end
+
+      if (r.request_state == 'active' || r.request_state == 'pending') && r.status == 'Ok' && r.process == true
```

Comment:
SQL for ```VmRetireRequest``` and ```ServiceRetireRequest``` are different since indexed columns ```source_id```  and ```source_type``` used for finding ```VmRetireRequest``` are not populated when ```ServiceRetireRequest``` (as well as ```OrchestrationStackRetireRequest```) created . 
I assumed that it would be much more approved request than not approved and It would require enumeration of all approved ```ServiceRetireRequest``` with extracting ```options[:src_id]``` from record in order  to do the same logic for ```ServiceRetirementRequest``` as for ```VmRetirementRequest``` and this operation could take a while on big table.

Also, when we discussed, we agreed that not approved request are main concern.
Please let me know if you think it is OK to do the same logic for service/stock retirement requests as for vm, and i will do it

Reply:
I mean if (based on observation on big clients) there are not  a lot of approved ```ServiceRetireRequest``` than we can do the same logic.

Reply:
Thanks @yrudman.  Yes, that's exactly right. My primary concern are the ```miq_requests``` that have not been approved. 

I know the data differs significantly in term of where/how to determine the source object, but I think we need the "active" check for ```Services``` and ```OrchestrationStacks```.  Without it, we might consider failed and/or finished requests as valid duplicates, which would prevent the request creation.

Reply:
I thought that all failed or finish requests would not have  ```:approval_state => "pending_approval"``` and changes in this PR do not  prevent request creation for failed or finished requests.

Is it possible for not approved request  (which has attribute ```:approval_state``` set to ```pending_approval``` ) be in failed or finished state ?

Reply:
very first draft of this PR was checking  "active" retirement request for service and stack, but during discussion we agreed that moving approval check into SQL and skipping analysis for approved (meaning active) request will improve response time and cover majority of possible cases

Reply:
Hi @yrudman,
The status check is important because there's no guarantee that all failed or finished requests have an ```:approval_state => "approved"```.

Reply:
Hi @tinaafitz 
I am not checking for ```:approval_state => "approved"``` but for ```:approval_state => "pending_approval"``` and only those not approved requests will be flagged as duplicate

OK, i think i know were my confusion coming from:
I was under impression that any processing of request (which would lead to failed or finished) will first changed ```:approval_state``` from ```"pending_approval"``` to something else;
and your point that  approval state could be "pending_approval" but request finished/failed, right ?

OK, i will check status for not approved requests

Reply:
Hi @yrudman, Yes, that's exactly right. My point is that the request could be in a "pending_approval" approval state, but be finished/failed. 

Could you add a check for all 3 object types?